### PR TITLE
[Substrait] [Python] [R] Upgrade Substrait and introduce function to export query plan as a substrait - JSON

### LINKED ISSUE #4098: Upgrade Substrait support
Hi, I've been using the [DuckDB API](https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/tests/extensions/substrait/test_substrait.py) to generate substrait proto bytes for an input query, in order to utilize Substrait plans infused with DuckDB optimizations for one of our internal projects.

1. It looks like the API supports Substrait 0.12, but the [latest Substrait](https://github.com/substrait-io/substrait/tree/main/proto/substrait) version (0.6) has had several breaking changes. We were wondering when the substrait version supported by the DuckDB API can be upgraded?

2. The API currently only generates proto bytes, and we're also curious if support can be added to convert the proto bytes to a json or protobuf file containing the DuckDB query plan in the [Substrait format](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto)?

Looking forward to your response, thank you very much!

### PR DESCRIPTION
**Pull request by pdet**
Fix: #4098 

This PR basically updates Substrait to its latest version, inlines a part of protobuf that is necessary to export serialized objects to JSON, and introduced functions on SQL,Python and R APIs to get the JSON format of substrait queries.

In SQL:
```sql
CREATE TABLE crossfit (exercise text,dificulty_level int);
CALL get_substrait_json('select count(exercise) as exercise from crossfit where dificulty_level <=5') 
```

In Python:
```python 
connection.execute('CREATE TABLE integers (i integer)')
json =  connection.get_substrait_json("select * from integers limit 5").fetchone()[0]

# OR
duckdb.default_connection.execute('CREATE TABLE integers (i integer)')
json = duckdb.get_substrait_json("select * from integers limit 5").fetchone()[0]

```

In R:
```r 
dbExecute(con, "CREATE TABLE integers (i INTEGER)")
json <- duckdb::duckdb_get_substrait_json(con, "select * from integers limit 5")
```

### CODE REVIEW

**Comment by pdet** [comment]
@gforsyth What is this used/necessary for again?

**Comment by gforsyth** [comment]
> @gforsyth What is this used/necessary for again?

If you build the python package from within `tools/pythonpkg` by running a `pip install .` then it seems to use the `substrait_config.py` file and ignores the CMakeLists

I was getting a bunch of missing symbol errors when trying to run the tests to try to replicate the error with Ibis.  After I added those files to `substrait_config.py` and rebuilt I can run the tests and they pass against both the current release of ibis-substrait and also against a branch with substrait upgraded to 0.8 (https://github.com/ibis-project/ibis-substrait/pull/278)

**Comment by pdet** [comment]
Ach so, that makes sense.

But it's this is a problem now, because Substraits is not bundled on the python client anymore, and it must be installed and loaded as an extension. This file should be removed and never used, I'll have a look on where it is being called in the python setup file tomorrow.

**Comment by gforsyth** [comment]
Ok, we just released ibis-substrait 2.9.3 and it's built on top of substrait 0.8.

If I build this branch locally and uncomment the ibis tpch tests they pass.

File: tools/pythonpkg/tests/extensions/substrait/test_ibis_tpch.py [hunk 1]
```diff
@@ -76,83 +76,83 @@ def execute(require, query_number):
     duck_con = TPCHBackend(duck_con=connection)
     tpch_execute_ibis_to_duck_query(duck_con,query_number)
 
-def test_query_substrait_01(require):
-    execute(require,1)
+# def test_query_substrait_01(require):
```
Comment (Mytherin) [inline tools/pythonpkg/tests/extensions/substrait/test_ibis_tpch.py:79]: Should this be commented out?

**Review by Mytherin — COMMENTED** [review]
LGTM, one comment

File: tools/pythonpkg/tests/extensions/substrait/test_ibis_tpch.py (hunk 1 above)
Comment (pdet) [reply tools/pythonpkg/tests/extensions/substrait/test_ibis_tpch.py:79]: Because of @gforsyth update yday, not anymore :-)

**Comment by gforsyth** [comment]
Wiped out `build/` and started in a fresh environment and this is working on my end:

Built with
```shell
BUILD_SUBSTRAIT_EXTENSION=1 BUILD_PYTHON=1 make
```

and test outputs and versions

```shell
(duckdb_dev) gil@stravaig ~/g/d/du/to/pythonpkg substrait_4098 🐚  pytest -k ibis
========================================== test session starts ===========================================
platform linux -- Python 3.9.10, pytest-7.1.0, pluggy-1.0.0
rootdir: /home/gil/github.com/duckdb/duckdb/tools/pythonpkg
collected 475 items / 453 deselected / 22 selected                                                       

tests/extensions/substrait/test_ibis_tpch.py .X.X..XXX.X.XXXXXX.XXX                                [100%]

============================= 7 passed, 453 deselected, 15 xpassed in 3.25s ==============================
(duckdb_dev) gil@stravaig ~/g/d/du/to/pythonpkg substrait_4098 🐚  ver duckdb
0.4.1-dev974
(duckdb_dev) gil@stravaig ~/g/d/du/to/pythonpkg substrait_4098 🐚  ver ibis
3.1.0
(duckdb_dev) gil@stravaig ~/g/d/du/to/pythonpkg substrait_4098 🐚  ver ibis_substrait
2.9.3
```

File: tools/pythonpkg/tests/extensions/substrait/test_substrait_json.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    if connection is None:
+        return
+
+    connection.execute('CREATE TABLE integers (i integer)')
+    json =  connection.get_substrait_json("select * from integers limit 5").fetchone()[0]
+    expected_result = '{"relations":[{"root":{"input":{"fetch":{"input":{"project":{"input":{"project":{"input":{"read":{"baseSchema":{"names":["i"]},"namedTable":{"names":["integers"]}}},"expressions":[{"selection":{"directReference":{"structField":{}}}}]}},"expressions":[{"selection":{"directReference":{"structField":{}}}}]}},"count":"5"}},"names":["i"]}}]}'
+    assert json == expected_result
+
+    # Test broken query
+    with pytest.raises(Exception, match="Table with name p does not exist!"):
```
Comment (Mause) [inline tools/pythonpkg/tests/extensions/substrait/test_substrait_json.py:15]: Any chance you could use a more specific exception type here instead?

**Review by Mause — CHANGES_REQUESTED** [review]

**Comment by pdet** [comment]
Alright, I think I figured out what the issue is.
In the latest protobuf version they deprecated the ```args()```in favor of ```.arguments()```  for the functions.

I think IBIS is still using the old model, which now breaks on duckdb since it expects these arguments.

e.g., scalar function from ibis:
```

(const substrait::Expression_ScalarFunction) $1 = {
  duckdb::google::protobuf::Message = {
    duckdb::google::protobuf::MessageLite = {
      _internal_metadata_ = (ptr_ = 0)
    }
  }
  args_ = {
    duckdb::google::protobuf::internal::RepeatedPtrFieldBase = {
      arena_ = nullptr
      current_size_ = 2
      total_size_ = 4
      rep_ = 0x0000600000c56490
    }
  }
  arguments_ = {
    duckdb::google::protobuf::internal::RepeatedPtrFieldBase = {
      arena_ = nullptr
      current_size_ = 0
      total_size_ = 0
      rep_ = nullptr
    }
  }
  output_type_ = 0x000060000022c800
  function_reference_ = 1
  _cached_size_ = {
    size_ = {
      Value = 0
    }
  }
}
```

For now I'm adding support to these on the consumer side again

**Comment by gforsyth** [comment]
> In the latest protobuf version they deprecated the `args()`in favor of `.arguments()` for the functions.
> 
> I think IBIS is still using the old model, which now breaks on duckdb since it expects these arguments.

Ahh, yeah, I bet that's it (we are still using `args`) -- I'll try to update that today.

**Comment by gforsyth** [comment]
Branch up using `arguments` instead of `args`: https://github.com/ibis-project/ibis-substrait/pull/289

**Pull request merged**